Panic when decoding json string

Hi! 
When I was trying to decode a json string into an interface, I got a panic.
Seems it happened when unescaping Unicode characters.
Here is an example:
``` go
package main

import (
	"bytes"
	"github.com/goccy/go-json"
)

func main() {
	a := `{"嗷嗷":["aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa\u55f7"]}`
	var o interface{}
	json.NewDecoder(bytes.NewBufferString(a)).Decode(&o)
}

```
Output:
```
panic: runtime error: slice bounds out of range [:504] with capacity 503

goroutine 1 [running]:
github.com/goccy/go-json/internal/decoder.decodeUnicodeRune(0xc00005a180, 0x0, 0x5c2900, 0x4ef360, 0x4e85a0, 0x7ff62a568858, 0xe000e000e000e)
	/home/abc/go/pkg/mod/github.com/goccy/go-json@v0.7.6/internal/decoder/string.go:107 +0x31d
github.com/goccy/go-json/internal/decoder.decodeUnicode(0xc00005a180, 0x0, 0xc000010310, 0x50a121, 0x7ff65123f9f0)
	/home/abc/go/pkg/mod/github.com/goccy/go-json@v0.7.6/internal/decoder/string.go:127 +0x4d
github.com/goccy/go-json/internal/decoder.decodeEscapeString(0xc00005a180, 0x0, 0x8, 0x18, 0x7ff651234108)
	/home/abc/go/pkg/mod/github.com/goccy/go-json@v0.7.6/internal/decoder/string.go:161 +0x405
github.com/goccy/go-json/internal/decoder.(*interfaceDecoder).decodeStreamEmptyInterface(0xc000064050, 0xc00005a180, 0x2, 0xc0000580a0, 0x0, 0x0)
	/home/abc/go/pkg/mod/github.com/goccy/go-json@v0.7.6/internal/decoder/interface.go:238 +0x21e
github.com/goccy/go-json/internal/decoder.(*interfaceDecoder).DecodeStream(0xc000064050, 0xc00005a180, 0x2, 0xc0000580a0, 0x18, 0x4e7140)
	/home/abc/go/pkg/mod/github.com/goccy/go-json@v0.7.6/internal/decoder/interface.go:318 +0x1bc
github.com/goccy/go-json/internal/decoder.(*sliceDecoder).DecodeStream(0xc0001260e0, 0xc00005a180, 0x2, 0xc00000c078, 0xc000124002, 0xc00006cb68)
	/home/abc/go/pkg/mod/github.com/goccy/go-json@v0.7.6/internal/decoder/slice.go:156 +0x556
github.com/goccy/go-json/internal/decoder.(*interfaceDecoder).decodeStreamEmptyInterface(0xc000064050, 0xc00005a180, 0x1, 0xc000010300, 0x0, 0x0)
	/home/abc/go/pkg/mod/github.com/goccy/go-json@v0.7.6/internal/decoder/interface.go:225 +0x5e5
github.com/goccy/go-json/internal/decoder.(*interfaceDecoder).DecodeStream(0xc000064050, 0xc00005a180, 0x1, 0xc000010300, 0x0, 0x0)
	/home/abc/go/pkg/mod/github.com/goccy/go-json@v0.7.6/internal/decoder/interface.go:318 +0x1bc
github.com/goccy/go-json/internal/decoder.(*mapDecoder).DecodeStream(0xc00005a240, 0xc00005a180, 0x1, 0xc000049d90, 0xc00006cd80, 0xc00006cd70)
	/home/abc/go/pkg/mod/github.com/goccy/go-json@v0.7.6/internal/decoder/map.go:94 +0x31e
github.com/goccy/go-json/internal/decoder.(*interfaceDecoder).decodeStreamEmptyInterface(0xc0000640a0, 0xc00005a180, 0x0, 0xc000010290, 0x0, 0x0)
	/home/abc/go/pkg/mod/github.com/goccy/go-json@v0.7.6/internal/decoder/interface.go:217 +0x6e5
github.com/goccy/go-json/internal/decoder.(*interfaceDecoder).DecodeStream(0xc0000640a0, 0xc00005a180, 0x0, 0xc000010290, 0x0, 0xc00006cee8)
	/home/abc/go/pkg/mod/github.com/goccy/go-json@v0.7.6/internal/decoder/interface.go:318 +0x1bc
github.com/goccy/go-json.(*Decoder).DecodeWithOption(0xc00006cf70, 0x4e4600, 0xc000010290, 0x0, 0x0, 0x0, 0x0, 0xc000038778)
	/home/abc/go/pkg/mod/github.com/goccy/go-json@v0.7.6/decode.go:202 +0x131
github.com/goccy/go-json.(*Decoder).Decode(...)
	/home/abc/go/pkg/mod/github.com/goccy/go-json@v0.7.6/decode.go:168
main.main()
	/home/abc/go/src/test/main.go:11 +0x1ce
```

--- a/internal/decoder/string.go
+++ b/internal/decoder/string.go
@@ -239,14 +239,15 @@ func stringBytes(s *Stream) ([]byte, error) {
 			fallthrough
 		default:
 			// multi bytes character
-			r, _ := utf8.DecodeRune(s.buf[cursor:])
-			b := []byte(string(r))
+			r, size := utf8.DecodeRune(s.buf[cursor:])
 			if r == utf8.RuneError {
-				s.buf = append(append(append([]byte{}, s.buf[:cursor]...), b...), s.buf[cursor+1:]...)
+				s.buf = append(append(append([]byte{}, s.buf[:cursor]...), runeErrBytes...), s.buf[cursor+1:]...)
+				cursor += runeErrBytesLen
+				s.length += runeErrBytesLen
 				_, _, p = s.stat()
+			} else {
+				cursor += int64(size)
 			}
-			cursor += int64(len(b))
-			s.length += int64(len(b))
 			continue
 		}
 		cursor++

--- a/internal/decoder/string.go
+++ b/internal/decoder/string.go
@@ -247,7 +247,7 @@ func stringBytes(s *Stream) ([]byte, error) {
 				}
 				goto ERROR
 			}
-      r, size := utf8.DecodeRune(s.buf[cursor:])
+			r, size := utf8.DecodeRune(s.buf[cursor:])
 			if r == utf8.RuneError {
 				s.buf = append(append(append([]byte{}, s.buf[:cursor]...), runeErrBytes...), s.buf[cursor+1:]...)
 				cursor += runeErrBytesLen

--- a/stream_test.go
+++ b/stream_test.go
@@ -512,7 +512,7 @@ func TestLongUTF8(t *testing.T) {
 		t.Fatalf("Unexpected error: %v", err)
 	}
 	if got != want {
-    t.Errorf("string %q; want = %q", got, want)
+		t.Errorf("string %q; want = %q", got, want)
 	}
 }
 

--- a/stream_test.go
+++ b/stream_test.go
@@ -14,7 +14,6 @@ import (
 	"net/http"
 	"net/http/httptest"
 	"reflect"
-
 	"strings"
 	"testing"
 
@@ -500,5 +499,18 @@ func TestGzipStreaming(t *testing.T) {
 	err = dec.Decode(&v)
 	if err != nil {
 		t.Fatalf("Unexpected error: %v", err)
+	}
+}
+
+func TestIssue278(t *testing.T) {
+	a := `{"嗷嗷":"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa\u55f7"}`
+	r := strings.NewReader(a)
+	var m map[string]string
+	if err := json.NewDecoder(r).Decode(&m); err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+	want := "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa\u55f7"
+	if got := m["嗷嗷"]; got != want {
+		t.Errorf("string %q; want = %q", got, want)
 	}
 }

